Upgrade to Python 3.9

allanlewis opened the pull request:
In order to use a version of Python with longer support, this commit upgrades the build to Python 3.9.

This is currently running in Sandbox.

Commit:
Dockerfile: Upgrade to Python 3.9

Python 3.7 will reach end-of-life in June 2023 [1]. In order to ensure we are
using a supported version of Python, this commit updates the build to use
Python 3.9, the latest supported by AWS Lambda.

This upgrades the underlying OS from Amazon Linux 1 to Amazon Linux 2, which
has a different set of base packages installed. By trial and error, it was
determined that we need to install some additional packages from the AL2 repos,
hence the additions to the `yumdownloader` command in the Dockerfile.

[1]: https://endoflife.date/python

Commit:
Dockerfile: Extract all downloaded RPMs

In order to avoid brittleness related to the naming of RPM files, this commit
replaces the explicit RPM extract commands with a loop through all of the
`.rpm` files. Before doing this, we ensure that some RPM files were actually
downloaded - otherwise the loop would exit without error.

Commit:
Dockerfile: [Fix] Actually fail when package to download doesn't exist

Commit 11f75b0 set the `skip_missing_names_on_install` option for
`yumdownloader` on the understanding that this would cause the command to fail
if any of the provides package names didn't exist. However, it appears that
this option doesn't work with `yumdownloader` so invalid package names were
_not_ causing the command to fail.

In order to work around this issue, this commit drops the redundant option and
adds some commands to the pipeline to make `yumdownloader` fail for missing
packages.

Implementation notes:

* `yumdownloader` prints "No Match for argument foo" (with that capitalisation)
when a package "foo" doesn't exist in the repos.

* `tee` is used to ensure that the output of `yumdownloader` is still logged;
otherwise the `grep` would elide it.

Commit:
upgrade_mime_check.py: Delete unused module

We always set `MIME_VALIDATION` to `no-validation`, which effectively disables
this code. As `python-magic` has system dependencies (`libmagic`) which can be
problematic on some platforms, this commit removes the MIME checking module and
its tests, and removes `python-magic` from the dependencies.

Commit:
Dockerfile: Install build dependencies separately from runtime dependencies

In order to simplify development of this Docker image, this commit splits the
installation of build-time dependencies from that of runtime dependencies. This
specifically allows the latter to be amended without reinstalling the former.

joshaw commented on Dockerfile:
```diff
@@ -23,12 +23,8 @@ RUN yum install -y --setopt skip_missing_names_on_install=False \
-    rpm2cpio gnutls*.rpm | cpio -idmv && \
-    rpm2cpio json-c*.rpm | cpio -idmv && \
-    rpm2cpio libtasn1*.rpm | cpio -idmv && \
+    { compgen -G "*.rpm" > /dev/null || { echo 'No RPMs downloaded!'; exit 1; }; } && \
+    for rpm in *.rpm; do rpm2cpio $rpm | cpio -idmv; done && \
```
Interesting approach to checking for the presence of files matching the pattern :thinking: 

We want to fail the build if _any_ of the rpms fail to be downloaded which is what your `tee | grep` dance does, so that should cover the case where no rpms are downloaded, right?. In which case, this might be simpler with `shopt -s nullglob` which just causes patterns that expand to nothing to be replaced by the empty string, rather than themselves ([docs](https://www.gnu.org/software/bash/manual/bash.html#The-Shopt-Builtin)).

joshaw replied:
Quote the `$rpm` variable.

joshaw commented on scripts/run-update-lambda:
```diff
@@ -26,4 +26,4 @@ docker run --rm \
   --cpus="${CPUS}" \
   --name="${NAME}" \
-  lambci/lambda:python3.7 update.lambda_handler
+  lambci/lambda:python3.9 update.lambda_handler
```
Not suggesting it should be changed here, but this is a very good reason to do a wider refactor of the AV scanner - this image hasn't been updated in 2 years.

Actually, having written that, I don't see a 3.9 version of that image at all
```
$ docker image pull lambci/lambda:python3.9
Error response from daemon: manifest for lambci/lambda:python3.9 not found: manifest unknown: manifest unknown
```

joshaw reviewed (dismissed):
Approved because I'm not sure of the impact/use of the `run-update-lambda` script - the updated version (3.9) of the image it uses doesn't appear to exist. But that could be an issue.

Commit:
clamav: [Fix] Don't unnecessarily prepend existing lib path

The environment variable `LD_LIBRARY_PATH` [1] sets paths to be searched for
libraries _in addition to_ the standard directories. It's therefore unnecessary
to include the standard directories when setting this variable. In order to
remove a dependency on `ld`, which isn't available in later versions of the
Lambda runtime, this commit removes the associated logic so that it simply sets
`LD_LIBRARY_PATH` to the _additional_ path we need.

[1]: https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html

allanlewis commented:
Closing until I've found a solution to a problem with the updater.
